deploy/helm/scf/templates/ingress.yaml

```diff
@@ -0,0 +1,82 @@
+{{- if .Values.features.ingress.enabled -}}
+---
+# The certificate and key for the TLS secret are passed through ingress.tls.crt and ingress.tls.key
+# respectively. If the operator does not provide these values at installation time, the TLS secret
+# will contain empty values. The standard behaviour for NGINX ingress controller is to provide a
+# fake certificate instead. It is useful only for testing and development. It is expected that for
+# production use the operator will provide these values.
+apiVersion: "v1"
+kind: "Secret"
+type: kubernetes.io/tls
+metadata:
+  name: "ingress-tls"
+  namespace: {{ .Release.Namespace | quote }}
+  labels:
+    app.kubernetes.io/component: "ingress-tls"
+    app.kubernetes.io/instance: {{ .Release.Name | quote }}
+    app.kubernetes.io/managed-by: {{ .Release.Service | quote }}
+    app.kubernetes.io/name: {{ default .Chart.Name .Values.nameOverride | trunc 63 | trimSuffix "-" | quote }}
+    app.kubernetes.io/version: {{ default .Chart.Version .Chart.AppVersion | quote }}
+    helm.sh/chart: {{ printf "%s-%s" .Chart.Name (.Chart.Version | replace "+" "_") | quote }}
+data:
+  tls.crt: {{ .Values.features.ingress.tls.crt | default "" | b64enc | quote }}
+  tls.key: {{ .Values.features.ingress.tls.key | default "" | b64enc | quote }}
+---
+# This ingress specifies routing and access for the cloud controller public SCF service, e.g.
+# "api.<domain>" and other services in that domain hierarchy.
+#
+# Note that the two other services (TCP routing and app SSH access) are not suitable to routing via
+# HTTP, and not specified. I.e. they are TCP, which is not supported by the current ingress
+# specification. Although specific controllers may do so in a custom manner.
+apiVersion: "extensions/v1beta1"
+kind: "Ingress"
+metadata:
+  name: {{ .Release.Name | quote }}
+  namespace: {{ .Release.Namespace | quote }}
+  annotations:
+    {{- if hasKey .Values.features.ingress.annotations "kubernetes.io/ingress.class" | not -}}
+      {{ $_ := set .Values.features.ingress.annotations "kubernetes.io/ingress.class" "nginx" }}
+    {{- end }}
```

Comment:
This all feels overly complicated. Could this be replaced by
```
kubernetes.io/ingress.class: {{ default "nginx" .Values.features.ingress.annotations | quote }}
```
And similar for the following clauses.

Alternatively, more questions:
  - Is `if haskey ... | not` the same as `if not (haskey ..)` ? If yes, I find the second more readable. Knowing from the outset that there is an inversion/negation. The current way gives me a surprise negation at the end, and I just managed to see that. Others may miss it entirely.
  - What does `$_ := set X Y Z` mean/do ?

Reply:
Using `default` here is hard because we want to use `{{ toYaml .Values.features.ingress.annotations | indent 4 }}` in line 52. For the bullets:
- I think it's the same. But indeed the way you suggested is more readable.
- There is a nice explanation in the sprig documentation here: http://masterminds.github.io/sprig/dicts.html.

Reply:
Improved in https://github.com/SUSE/scf/pull/2708/commits/4dedc618264fdbe13e0074f907b2f354120464b9.

Reply:
I still do not understand why the `toYaml` and the dynamic dict would be needed ...

:bulb: Ah. I might see it now.

I tried to rewrite this fully to defaults, and then noticed that the last element in the key I would use would stuff like "kubernetes.io/ingress.class", which might not be writable as dot notation.

This also triggered the realization that the user is maybe allowed to write __any__ annotation they like in that section. And the `toYaml` will then format and insert that properly. 

And thus it also becomes clear why `default` is not usable for the annotations which are required. If the user specifies nothing then default inserts something. No clash with `toYaml`.

But when the user does provide their own setting we will get it inserted twice, once from the `default and then again from the `toYaml`.

So you are checking explicitly, and dynamically extend the underlying dict instead, and everything goes through `toYaml`.

If that is correct, I would really like to have few comment sentences in the template which explain the basics of this for any future dev.

The main :bulb: was the realization that the `annotation:` key in the values is not a fixed set of things, but completely under user control.

Reply:
You are correct. Added the comment in https://github.com/SUSE/scf/pull/2708/commits/c75227ae676c2a4cc649f6443e5d65eb93102e5d.